yt_dlp/extractor/niconico.py

```diff
@@ -33,13 +32,72 @@
     urlencode_postdata,
     urljoin,
 )
+from ..utils.traversal import find_element, traverse_obj
 
 
-class NiconicoIE(InfoExtractor):
+class NiconicoBaseIE(InfoExtractor):
+    _GEO_BYPASS = False
+    _GEO_COUNTRIES = ['JP']
+    _LOGIN_BASE = 'https://account.nicovideo.jp'
+    _NETRC_MACHINE = 'niconico'
+
+    def raise_login_error(self, error, default, expected=True):
+        raise ExtractorError(f'Unable to login: {error or default}', expected=expected)
+
+    def _perform_login(self, username, password):
+        if self._get_cookies('https://www.nicovideo.jp').get('user_session'):
+            return
+
+        self._request_webpage(
+            f'{self._LOGIN_BASE}/login', None, 'Requesting session cookies')
+        webpage = self._download_webpage(
+            f'{self._LOGIN_BASE}/login/redirector', None,
+            'Logging in', headers={
+                'Content-Type': 'application/x-www-form-urlencoded',
+                'Referer': f'{self._LOGIN_BASE}/login',
+            }, data=urlencode_postdata({
+                'mail_tel': username,
+                'password': password,
+            }),
+        )
+
+        if traverse_obj(self._search_json(
+            r'NicoGoogleTagManagerDataLayer\s*=\s*\[',
+            webpage, 'data layer', None, fatal=False,
+        ), ('user', 'login_status', {lambda x: x == 'login'})):
+            return
+
+        if err_msg := traverse_obj(webpage, (
+            {find_element(cls='notice error')}, {find_element(cls='notice__text')}, {clean_html},
+        )):
+            self.raise_login_error(err_msg, 'Invalid username or password')
+        elif 'oneTimePw' in webpage:
+            post_url = self._search_regex(
+                r'<form[^>]+action=(["\'])(?P<url>.+?)\1', webpage, 'post url', group='url')
+            mfa, urlh = self._download_webpage_handle(
+                urljoin(self._LOGIN_BASE, post_url), None,
+                'Performing MFA', 'Unable to complete MFA', headers={
+                    'Content-Type': 'application/x-www-form-urlencoded',
+                }, data=urlencode_postdata({
+                    'otp': self._get_tfa_info('6 digit number shown on app'),
+                }),
+            )
+
+            if 'error-code' in parse_qs(urlh.url):
+                err_msg = traverse_obj(self._download_webpage(urlh.url, None), (
+                    {find_element(cls='pageMainMsg')}, {clean_html}))
```

Comment:
Why do we need to download the webpage again? `urlh.url` is the current location of what was read as `mfa`. Can't we just do this?
```suggestion
                err_msg = traverse_obj(mfa, ({find_element(cls='pageMainMsg')}, {clean_html}))
```

Reply:
it was my mistake, I could simply get it from `mfa`